src/opnsense/mvc/app/controllers/OPNsense/Base/ApiMutableTableModelControllerBase.php

```diff
@@ -0,0 +1,183 @@
+ *    AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ *    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+namespace OPNsense\Base;
+
+/**
+ * Class ApiMutableTableModelControllerBase, inherit this class to implement
+ * an API that exposes a model for tables.
+ * @package OPNsense\Base
+ */
+abstract class ApiMutableTableModelControllerBase extends ApiControllerBase
+{
+    // FIXME What about validation?
+
+    static protected $modelPathPrefix = '';
+    private function getNodes() {
+        $ref = static::$modelPathPrefix
+             + static::$internalModelName;
+        return $this->getModel()->getNodeByReference($ref);
+    }
+    private function getNodeByUUID($uuid) {
+        return $this->getNodes()->$uuid;
+    }
+
+    /**
+     * retrieve item or return defaults
+     * @param $uuid item unique id
+     * @return array
+     */
+    public function getItemAction($uuid = null)
+    {
+        $mdl = $this->getModel();
+        if ($uuid != null) {
+            $node = $this->getNodeByUUID($uuid);
+            if ($node != null) {
+                // return node
+                return array(static::$internalModelName => $node->getNodes());
+            }
+        } else {
+            // generate new node, but don't save to disc
+            $node = getNodes()->add();
+            return array(static::$internalModelName => $node->getNodes());
+        }
+        return array();
+    }
+
+    /**
+     * update item with given properties
+     * @param $uuid item unique id
+     * @return array
+     */
+    public function setItemAction($uuid)
+    {
+        if ($this->request->isPost() && $this->request->hasPost(static::$internalModelName)) {
+            $mdl = $this->getModel();
+            if ($uuid != null) {
+                $node = $this->getNodeByUUID($uuid);
+                if ($node != null) {
+                    $node->setNodes($this->request->getPost(static::$internalModelName));
+                    return $this->save($mdl, $node, static::$internalModelName);
+                }
+            }
+        }
+        return array("result"=>"failed");
+    }
+
+    /**
+     * add new item and set with attributes from post
+     * @return array
+     */
+    public function addItemAction()
+    {
+        $result = array("result"=>"failed");
+        if ($this->request->isPost() && $this->request->hasPost(static::$internalModelName))) {
+            $mdl = $this->getModel();
+            // FIXME Is this correct?
+            $node = $this->getNodes()->add();
```

Comment:
Why uppercase? Isn't this the function we're trying to call? https://github.com/opnsense/core/blob/master/src/opnsense/mvc/app/models/OPNsense/Base/FieldTypes/ArrayField.php#L76